pyro/contrib/scoped.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import absolute_import, division, print_function
+
+import functools
+import uuid
+
+import pyro
+import pyro.infer
+import pyro.contrib.named as named
+
+
+def _lift(fn, _namespace=None, _latent=None):
+    @functools.wraps(fn)
+    def decorated(*args, **kwargs):
+        latent = named.Object(_namespace) if _latent is None else _latent
+        return fn(latent, *args, **kwargs)
+    return decorated
+
+
+@functools.wraps(pyro.condition)
+def condition(fn, data=None, data_fn=None):
+    def partial_condition(latent, *args, **kwargs):
+        if data is not None:
+            lifted_data = {}
+            for key, value in data.items():
+                lifted_data["{}.{}".format(latent._name, key)] = value
+        else:
+            site_collector = named.Object(latent._name)
+            data_fn(site_collector)
+            lifted_data = site_collector.visit(lambda name, val, acc: acc.setdefault(name, val),
```

Comment:
And if you replace `.visit()` with the `.items()` suggested above, this simplifies to
```py
lifted_data = dict(site_collector.items())
```